lib/selfNodeId.ts

```diff
@@ -0,0 +1,42 @@
+import { DirectiveLocation } from 'graphql';
+import { ValidationError } from 'apollo-server-errors';
+
+import ValidateDirectiveVisitor, {
+  ValidateFunction,
+} from './ValidateDirectiveVisitor';
+
+type ToNodeId = (entityName: string, id: string) => string;
+
+export type SelfNodeIdContext<TContext extends object = object> = {
+  toNodeId: ToNodeId;
+};
+
+export default class SelfNodeIdDirective<
+  TContext extends SelfNodeIdContext
+> extends ValidateDirectiveVisitor<{}, SelfNodeIdContext> {
+  public getValidationForArgs():
+    | ValidateFunction<SelfNodeIdContext>
+    | undefined {
+    return (value: unknown, _, container, { toNodeId }): string => {
+      if (typeof value !== 'string') {
+        throw new ValidationError(
+          `${this.name} directive only works on strings`,
+        );
+      }
+      return toNodeId(container.name, value);
+    };
+  }
+
+  public static readonly config: typeof ValidateDirectiveVisitor['config'] = {
+    description: 'ensures that the ID is converted to a global ID',
+    locations: [DirectiveLocation.FIELD_DEFINITION],
```

Comment:
maybe we could also support `OBJECT`, in this case we'd look for `fields.id` and annotate it? it's a nice usage helper.